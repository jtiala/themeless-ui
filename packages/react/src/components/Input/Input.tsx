import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { InputHTMLAttributes } from "react";

type HTMLInputProps = InputHTMLAttributes<HTMLInputElement>;

export type InputProps = {
  /**
   * Type of the input.
   */
  type?: HTMLInputProps["type"];

  /**
   * Name of the input.
   */
  name?: HTMLInputProps["name"];

  /**
   * Prevent the user from interacting with the input.
   */
  disabled?: HTMLInputProps["disabled"];

  /**
   * Indicates that the user should not be able to edit the value of the input.
   */
  readOnly?: HTMLInputProps["readOnly"];

  /**
   * Indicates that the user must specify a value for the input before the owning form can be submitted.
   */
  required?: HTMLInputProps["required"];

  /**
   * Brief hint for the user as to what kind of information is expected in the input.
   */
  placeholder?: HTMLInputProps["placeholder"];

  /**
   * The value of the input.
   * Use only when using the input as controlled component (state handled by React).
   */
  value?: HTMLInputProps["value"];

  /**
   * Callback to call when the input's value is modified.
   */
  onChange?: HTMLInputProps["onChange"];

  /**
   * Callback to call when the input is clicked.
   */
  onClick?: HTMLInputProps["onClick"];

  /**
   * Callback to call when the input receives focus.
   */
  onFocus?: HTMLInputProps["onFocus"];

  /**
   * Callback to call when the input loses focus.
   */
  onBlur?: HTMLInputProps["onBlur"];
} & CommonComponentProps;

const className = cn("input");

/**
 * Display an input.
 */
export function Input({
  type = "text",
  name,
  disabled,
  readOnly,
  required,
  placeholder,
  value,
  onChange,
  onClick,
  onFocus,
  onBlur,
  id,
  testId,
}: InputProps) {
  return (
    <input
      type={type}
      name={name}
      disabled={disabled}
      readOnly={readOnly}
      required={required}
      placeholder={placeholder}
      value={value}
      onChange={onChange}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      className={className}
      id={id}
      data-testid={testId}
    />
  );
}
