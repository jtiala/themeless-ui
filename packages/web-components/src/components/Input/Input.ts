import style from "@themeless-ui/style/dist/input.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import {
  TUIBlurEvent,
  TUIChangeEvent,
  TUIClickEvent,
  TUIFocusEvent,
  TUIMouseEnterEvent,
  TUIMouseLeaveEvent,
  TUITouchEndEvent,
  TUITouchStartEvent,
} from "../../events";
import { TUIComponent } from "../TUIComponent";

const className = cn("input");

@customElement("tui-input")
export class Input extends TUIComponent {
  /**
   * Type of the input.
   *
   * @default text
   */
  @property()
  type?: HTMLInputElement["type"];

  /**
   * Name of the input.
   */
  @property()
  name?: HTMLInputElement["name"];

  /**
   * Prevent the user from interacting with the input.
   */
  @property({ type: Boolean })
  disabled?: HTMLInputElement["disabled"];

  /**
   * Indicates that the user should not be able to edit the value of the input.
   */
  @property({ type: Boolean })
  readOnly?: HTMLInputElement["readOnly"];

  /**
   * Indicates that the user must specify a value for the input before the owning form can be submitted.
   */
  @property({ type: Boolean })
  required?: HTMLInputElement["required"];

  /**
   * Brief hint for the user as to what kind of information is expected in the input.
   */
  @property()
  placeholder?: HTMLInputElement["placeholder"];

  /**
   * The value of the input.
   * Use only when using the input as controlled component (state handled by React).
   */
  @property()
  value?: HTMLInputElement["value"];

  private _handleChange() {
    this.dispatchEvent(new TUIChangeEvent());
  }

  private _handleClick() {
    this.dispatchEvent(new TUIClickEvent());
  }

  private _handleFocus() {
    this.dispatchEvent(new TUIFocusEvent());
  }
  private _handleBlur() {
    this.dispatchEvent(new TUIBlurEvent());
  }

  private _handleMouseEnter() {
    this.dispatchEvent(new TUIMouseEnterEvent());
  }

  private _handleMouseLeave() {
    this.dispatchEvent(new TUIMouseLeaveEvent());
  }

  private _handleTouchStart() {
    this.dispatchEvent(new TUITouchStartEvent());
  }

  private _handleTouchEnd() {
    this.dispatchEvent(new TUITouchEndEvent());
  }

  static styles = unsafeCSS(style);

  constructor() {
    super();

    this.type = "text";
  }

  override render() {
    return html`
      <input
        type="${this.type}"
        name="${this.name || nothing}"
        placeholder="${this.placeholder || nothing}"
        value="${this.value || nothing}"
        ?disabled="${this.disabled}"
        ?readOnly="${this.readOnly}"
        ?required="${this.required}"
        @change="${this._handleChange}"
        @click="${this._handleClick}"
        @focus="${this._handleFocus}"
        @blur="${this._handleBlur}"
        @mouseEnter="${this._handleMouseEnter}"
        @mouseLeave="${this._handleMouseLeave}"
        @touchStart="${this._handleTouchStart}"
        @touchEnd="${this._handleTouchEnd}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      />
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-input": Input;
  }
}
