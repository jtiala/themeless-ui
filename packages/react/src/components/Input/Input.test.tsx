import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
  userEvent,
  vi,
} from "../../../test";
import { Input } from "./Input";

const onChange = vi.fn();
const onClick = vi.fn();
const onFocus = vi.fn();
const onBlur = vi.fn();

const defaultInput = (
  <Input
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="input"
  />
);

const disabledInput = (
  <Input
    disabled
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="input"
  />
);

const readOnlyInput = (
  <Input
    readOnly
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="input"
  />
);

describe("Input", async () => {
  it("should render the input", () => {
    expect(componentToJson(renderer.create(defaultInput))).toMatchSnapshot();

    render(defaultInput);
    const input = screen.getByTestId("input");

    expect(input).toBeInTheDocument();
  });

  it("should call the callbacks", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(defaultInput);
    const input = screen.getByTestId("input");

    expect(input).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(input);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(1);
    expect(onFocus.mock.calls.length).toBe(1);

    input.blur();
    expect(onBlur.mock.calls.length).toBe(1);

    input.focus();
    expect(onFocus.mock.calls.length).toBe(2);

    await userEvent.type(input, "test");
    expect(onChange.mock.calls.length).toBe(4);
    expect(input).toHaveValue("test");
  });

  it("shouldn't be possible to access a disabled input", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(disabledInput);
    const input = screen.getByTestId("input");

    expect(input).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(input);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);

    input.blur();
    expect(onBlur.mock.calls.length).toBe(0);

    input.focus();
    expect(onFocus.mock.calls.length).toBe(0);

    await userEvent.type(input, "test");
    expect(onChange.mock.calls.length).toBe(0);
  });

  it("shouldn't be possible to change the value of a read only input", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(readOnlyInput);
    const input = screen.getByTestId("input");

    expect(input).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(input);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(1);
    expect(onFocus.mock.calls.length).toBe(1);

    input.blur();
    expect(onBlur.mock.calls.length).toBe(1);

    input.focus();
    expect(onFocus.mock.calls.length).toBe(2);

    await userEvent.type(input, "test");
    expect(onChange.mock.calls.length).toBe(0);
  });
});
