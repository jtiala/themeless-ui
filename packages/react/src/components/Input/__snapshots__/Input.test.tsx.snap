// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should render the input 1`] = `
<input
  className="tui__input"
  data-testid="input"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  type="text"
/>
`;
